app/src/main/java/app/revanced/manager/domain/repository/PatchBundleRepository.kt

```diff
@@ -304,13 +304,77 @@ class PatchBundleRepository(
         doReload()
     }
 
-    suspend fun createRemote(url: String, autoUpdate: Boolean) =
+    /**
+     * Creates a new remote patch bundle source.
+     * Validates and downloads the patch bundle before saving to ensure the URL is valid.
+     *
+     * @throws InvalidBundleSourceException if the URL is invalid or the download fails
+     */
+    suspend fun createRemote(url: String, autoUpdate: Boolean) {
+        // Validate URL format first
+        val source = try {
+            SourceInfo.from(url)
+        } catch (e: Exception) {
+            throw InvalidBundleSourceException(
+                app.getString(R.string.source_invalid_url),
+                e
+            )
+        }
+
+        // Ensure it's a remote source
+        if (source !is SourceInfo.Remote) {
+            throw InvalidBundleSourceException(
+                app.getString(R.string.source_invalid_url)
+            )
+        }
```

Comment:
Move this to the caller and make the function accept a `SourceInfo.Remote` instead.